### ISSUE TITLE
[Feature Request] - Only take photos of failed tests

### ISSUE THREAD

**Issue by jaredsburrows** [square/spoon#489]
Option to only take photos of failed tests. Similar to Junit reports that list the test failures only.

**Comment by JakeWharton** [comment]
Spoon doesn't take photos. Your tests do! You should be able to write a
rule or base class that only captures a screenshot when the test is failing.

**Comment by jaredsburrows** [comment]
@JakeWharton You are right but Spoon will still generate the html reports around the missing photos. Ideally it would be nice if spoon would only show/generate reports with screenshots based on the failed tests only.

**Comment by JakeWharton** [comment]
That is orthogonal to the feature being requested in this issue. If you
want an altered report then you can use the JSON to generate your own.

**Comment by jaredsburrows** [comment]
@JakeWharton Sorry for not being clear initially.

Example test:

```
@Test
fun testHomeScreen() {
 startApp();     // initial screen shot
 screenshot();   // screenshot 1
 doThing();      // interaction
 screenshot();   // after screenshot
}
```

Let's say this passes, I ideally do not want to keep both of those screen shots. If it fails during `doThing()`, the screenshots should be kept and the reports be generated accordingly.

This way, the reports can be smaller. Maybe this issue should be brought up in one of the gradle plugins? Such as https://github.com/stanfy/spoon-gradle-plugin?

**Comment by JakeWharton** [comment]
Spoon was designed to keep screenshots for the success case so you can see
your app in various configurations. It sounds like you want the opposite
which might indicate that it's not the right tool.

**Comment by albgz** [comment]
You can make your own listener for taking screenshots on failures and then run spoon with -e YourListenerClass.

` */
class SpoonTestFailureListener : RunListener() {
    private val NAME_SEPARATOR = "_"
    internal var desc: Description? = null

    override fun testStarted(description: Description?) {
        desc = description
    }

    @Throws(IOException::class)
    override fun testFailure(failure: Failure?) {
        val screenshot_file = Environment.getExternalStorageDirectory().absolutePath +
                "/app_spoon-screenshots/"+ desc!!.className + "/" +
                desc!!.methodName + "/"
        Helper.createDirIfNotExists(File(screenshot_file))
        val screenshotName = File(screenshot_file + System.currentTimeMillis() + NAME_SEPARATOR + "SpoonTestFailure" + EXTENSION)
        val device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
        device.takeScreenshot(screenshotName)
    }

    companion object {
        private val EXTENSION = ".png"
    }
}`
The Helper.createDirIfNotExists is a method to create the directory for the screenshots. Also be sure to clean the /sdcard/app_spoon-screenshots folder before running the tests.
P.S: Sorry for the messy kotlin code

**Comment by jaredsburrows** [comment]
@albgz Can you post in a gist or update your post with the whole implementation?

**Comment by albgz** [comment]
@jaredsburrows Here it is: https://gist.github.com/anonymous/48309d3f3323e945d729ec6b5321aaa1

**Comment by jaredsburrows** [comment]
I guess my only question is, will spoon generate the reports still with these new screenshots?

**Comment by albgz** [comment]
Yes, the screenshots will be automatically embedded by spoon in the reports under the failing tests.
